week3/homework/src/actions.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+const fs = require('fs');
+const uuid = require('uuid');
+
+function stringifyAndWrite(arr) {
+  fs.writeFileSync('./data.json', JSON.stringify(arr));
+}
+
+function readAndParseData() {
+  const string = fs.readFileSync('./data.json', 'utf8');
+  return JSON.parse(string);
+}
+
+function readF() {
+  return fs.readFileSync('./data.json', 'utf8');
+}
+
+function addToList(toDo) {
+  const toDoObject = {
+    task: toDo,
+    id: uuid(),
+    done: false,
+  };
+  const toDosArray = readAndParseData();
+  toDosArray.push(toDoObject);
+  stringifyAndWrite(toDosArray);
+}
+
+function findIndex(id, arr) {
+  return arr.findIndex(obj => obj.id === id);
+}
+
+function removeFromList(toDoId) {
+  const toDosArray = readAndParseData();
+  const index = findIndex(toDoId, toDosArray);
+  if (index === -1) {
+    return 'The list has no corresponding "to-do element"';
+  } else {
+    toDosArray.splice(index, 1);
+    stringifyAndWrite(toDosArray);
+  }
+}
+
+function readSingleToDo(toDoId) {
+  const toDosArray = readAndParseData();
+  const index = findIndex(toDoId, toDosArray);
+  if (index === -1) {
+    return 'The list has no corresponding "to-do element"';
+  } else {
+    return JSON.stringify(toDosArray[index]);
+  }
+}
+
+function markDone(toDoId) {
+  const toDosArray = readAndParseData();
+  const index = findIndex(toDoId, toDosArray);
+  if (index === -1) {
+    return 'The list has no corresponding "to-do element"';
+  } else {
+    if (toDosArray[index].done === true) {
+      return 'The task is already done.';
+    } else {
+      toDosArray[index].done = true;
+      stringifyAndWrite(toDosArray);
+    }
+  }
+}
+
+function markUnDone(toDoId) {
```

Comment:
In general in programming if you have `markAsTrue()` and `markAsFlase()` you could replace these two functions with one "toggle" function: you are changing the state to the opposite state. In this specific case, you could do something like:
```
function toggleDone(toDoId) {
  const toDosArray = readAndParseData();
  const index = findIndex(toDoId, toDosArray);
  toDosArray[index].done = !toDosArray[index].done;
}
```

Reply:
This is much easier and straightforward! 👍 @J05J4